### PR TITLE
Tobin gh915 invalid cb fb image bindings

### LINKED ISSUE #915: Crash on ValidateImageMemoryIsValid
_getImageNode(dev_data, fb_info.image)_ is returning null, and then _ValidateImageMemoryIsValid()_ will try to dereference the nullptr on _if (image_node->mem == MEMTRACKER_SWAP_CHAIN_IMAGE_KEY)_. The image got deleted before this submit. Callstack:
>	VkLayer_core_validation.dll!core_validation::ValidateImageMemoryIsValid(dev_data=0x00000000214bc490, image_node=0x0000000000000000, functionName=0x000007fec4f619c8) Line 459	C++
 	VkLayer_core_validation.dll!core_validation::CmdBeginRenderPass::__l24::<lambda>() Line 9994	C++
 	VkLayer_core_validation.dll!std::_Callable_obj<bool <lambda>(void),0>::_ApplyX<bool>() Line 284	C++
 	VkLayer_core_validation.dll!std::_Func_impl<std::_Callable_obj<bool <lambda>(void),0>,std::allocator<std::_Func_class<bool> >,bool>::_Do_call() Line 229	C++
 	VkLayer_core_validation.dll!std::_Func_class<bool>::operator()() Line 316	C++
 	VkLayer_core_validation.dll!core_validation::QueueSubmit(queue=0x000000001bf9ae60, submitCount=1, pSubmits=0x000000003122d250, fence=0x0000000000002d11) Line 4917	C++
 	VkLayer_object_tracker.dll!object_tracker::QueueSubmit(queue=0x000000001bf9ae60, submitCount=1, pSubmits=0x000000003122d250, fence=0x0000000000002d11) Line 635	C++
 	VkLayer_parameter_validation.dll!parameter_validation::QueueSubmit(queue=0x000000001bf9ae60, submitCount=1, pSubmits=0x000000003122d250, fence=0x0000000000002d11) Line 1802	C++
 	VkLayer_threading.dll!threading::QueueSubmit(queue=0x000000001bf9ae60, submitCount=1, pSubmits=0x000000003122d250, fence=0x0000000000002d11) Line 129	C++
 	VkLayer_api_dump.dll!vkQueueSubmit(queue=0x000000001bf9ae60, submitCount=1, pSubmits=0x000000003122d250, fence=0x0000000000002d11) Line 582	C++
 	UE4Editor-VulkanRHI-Win64-Debug.dll!FVulkanQueue::Submit(CmdBuffer=0x0000000017e3bcd0, WaitSemaphore=0x00000000410f1600, WaitStageFlags=8192, SignalSemaphore=0x0000000000000000) Line 70	C++

### PR DESCRIPTION
**Pull request by tobine**
Fixes #915 and includes some other minor clean-up as well as a test

Main fix is to bind images that are in framebuffer attachments to command buffers along with the framebuffer itself. Test hits this case and verifies that command buffer goes invalid.

Added early exit case for invalid command buffer at QueueSubmit time as null objects can then crash on subsequent queued function calls for memory set/validate if objects were destroyed and ptrs are null.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -4924,6 +4924,9 @@ QueueSubmit(VkQueue queue, uint32_t submitCount, const VkSubmitInfo *pSubmits, V
                 pCBNode->submitCount++; // increment submit count
                 skip_call |= validatePrimaryCommandBufferState(dev_data, pCBNode);
                 skip_call |= validateQueueFamilyIndices(dev_data, pCBNode, queue);
+                // Potential early exit here as bad object state may crash in delayed function calls
+                if (skip_call)
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: These delayed function calls are generally pretty murky wrt side effects. Avoiding the obvious crash seems like a good plan though.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -453,7 +453,7 @@ static const char *object_type_to_string(VkDebugReportObjectTypeEXT type) {
     case VK_DEBUG_REPORT_OBJECT_TYPE_DESCRIPTOR_SET_EXT:
         return "descriptor set";
     case VK_DEBUG_REPORT_OBJECT_TYPE_FRAMEBUFFER_EXT:
-        return "buffer";
+        return "framebuffer";
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Oops. That would have been a bit confusing. :+1:

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -6666,8 +6666,8 @@ BeginCommandBuffer(VkCommandBuffer commandBuffer, const VkCommandBufferBeginInfo
     layer_data *dev_data = get_my_data_ptr(get_dispatch_key(commandBuffer), layer_data_map);
     std::unique_lock<std::mutex> lock(global_lock);
     // Validate command buffer level
-    GLOBAL_CB_NODE *pCB = getCBNode(dev_data, commandBuffer);
-    if (pCB) {
+    GLOBAL_CB_NODE *cb_node = getCBNode(dev_data, commandBuffer);
+    if (cb_node) {
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Generally not thrilled about ever including _node, it's about as meaningful as _object... but OK.

File: tests/layer_validation_tests.cpp [hunk 4]
```diff
@@ -7613,6 +7613,155 @@ TEST_F(VkLayerTest, InvalidCmdBufferImageDestroyed) {
...
+
+    VkAttachmentDescription attachment = {
+        0,
+        VK_FORMAT_B8G8R8A8_UNORM,
+        VK_SAMPLE_COUNT_1_BIT,
+        VK_ATTACHMENT_LOAD_OP_DONT_CARE,
+        VK_ATTACHMENT_STORE_OP_STORE,
+        VK_ATTACHMENT_LOAD_OP_DONT_CARE,
+        VK_ATTACHMENT_STORE_OP_DONT_CARE,
+        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: `initialLayout` here is `COLOR_ATTACHMENT_OPTIMAL`, but the image you use it with is in `UNDEFINED` at the beginning of the render pass instance. Just change the layout here to `UNDEFINED` ?

File: tests/layer_validation_tests.cpp [hunk 5]
```diff
@@ -7613,6 +7613,155 @@ TEST_F(VkLayerTest, InvalidCmdBufferImageDestroyed) {
...
+
+    VkRenderPassBeginInfo rpbi = {VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO,
+                                  nullptr,
+                                  rp,
+                                  fb,
+                                  {{0, 0}, {32, 32}},
+                                  0,
+                                  nullptr};
+    BeginCommandBuffer();
+    vkCmdBeginRenderPass(m_commandBuffer->handle(), &rpbi,
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: UB ignoring error here due to layout problem

File: tests/layer_validation_tests.cpp [hunk 6]
```diff
@@ -7613,6 +7613,155 @@ TEST_F(VkLayerTest, InvalidCmdBufferImageDestroyed) {
...
+    VkAttachmentReference att_ref = {0,
+                                     VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL};
+
+    VkSubpassDescription subpass = {0,       VK_PIPELINE_BIND_POINT_GRAPHICS,
+                                    0,       nullptr,
+                                    0,       nullptr,
+                                    nullptr, &att_ref,
+                                    0,       nullptr};
+
+    VkSubpassDependency dep = {0,
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Not needed

File: tests/layer_validation_tests.cpp [hunk 7]
```diff
@@ -7613,6 +7613,155 @@ TEST_F(VkLayerTest, InvalidCmdBufferImageDestroyed) {
...
+                               0,
+                               VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
+                               VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
+                               VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT,
+                               VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT,
+                               VK_DEPENDENCY_BY_REGION_BIT};
+
+    VkRenderPassCreateInfo rpci = {VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO,
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Maybe don't even need the explicit renderpass at all, as long as your test framebuffer is compatible with the framework's one?

File: tests/layer_validation_tests.cpp [hunk 8]
```diff
@@ -7613,6 +7613,155 @@ TEST_F(VkLayerTest, InvalidCmdBufferImageDestroyed) {
...
+                                   &subpass,
+                                   1,
+                                   &dep};
+
+    VkResult err;
+    VkRenderPass rp;
+    err = vkCreateRenderPass(m_device->device(), &rpci, nullptr, &rp);
+    ASSERT_VK_SUCCESS(err);
+
+    VkImageCreateInfo image_ci = {};
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Use VkImageObj ? (you'd need to arrange for it to be destroyed at the right point, but that shouldn't be too bad...)

File: tests/layer_validation_tests.cpp [hunk 9]
```diff
@@ -7613,6 +7613,155 @@ TEST_F(VkLayerTest, InvalidCmdBufferImageDestroyed) {
...
+        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
+        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL};
+
+    VkAttachmentReference att_ref = {0,
+                                     VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL};
+
+    VkSubpassDescription subpass = {0,       VK_PIPELINE_BIND_POINT_GRAPHICS,
+                                    0,       nullptr,
+                                    0,       nullptr,
+                                    nullptr, &att_ref,
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: `att_ref` is being specified as the depth/stencil attachment for the subpass here, but it's a color attachment...

**Comment by chrisforbes** [comment]
Mostly :+1:, the test is wobbly though.

File: layers/core_validation.cpp (hunk 1 above)
Comment (mark-lunarg) [inline layers/core_validation.cpp]: +1, and +1.  They're fun to debug, too.

**Comment by mark-lunarg** [comment]
Heh, me giving a :+1: after @chrisforbes has taken a look is pretty pointless, but it makes me feel useful.

File: tests/layer_validation_tests.cpp (hunk 7 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: Yeah, good thought, I updated to this.

File: tests/layer_validation_tests.cpp (hunk 6 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: dead

File: tests/layer_validation_tests.cpp (hunk 4 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: Fo sho

File: layers/core_validation.cpp (hunk 3 above)
Comment (tobine) [inline layers/core_validation.cpp]: Yeah, we can overhaul everything to "state" at some point. This was more about lowercase in compliance with coding standard and we've already got cb_node lots of places.

File: tests/layer_validation_tests.cpp (hunk 8 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: Yeah, I looked at either scoping it or adding an early destroy method to the class but hit a couple roadblocks and got annoyed after about 10min so went this route.

**Comment by tobine** [comment]
Force pushed update with greatly simplified and improved test based on @chrisforbes expert guidance

**Comment by tobine** [comment]
cherry-picked and pushed to master

**Pull request closed without merging**